### PR TITLE
Allocation and performance benchmark tools

### PR DESCRIPTION
**Pull request by ktoso**
This introduces two benchmark suite tools.
They're based on existing tools in Swift or SSWG projects: Swift (the SwiftBenchmarkTools) itself and Swift NIO (allocation tests in IntegrationTests/) respectively.

_It's not an in depth analysis at all, nor meant as a full benchmark right now -- but this is to import the benchmark infra for when we'll need it._

----

The integration tests allow running and seeing allocations made:

```
./IntegrationTests/run-tests.sh  -v
```

Just for reference, the present day context gives (on mac):

```
test_002_pass_around_static_strings_large.total_allocations: 4
test_002_pass_around_static_strings_large.total_allocated_bytes: 528
test_002_pass_around_static_strings_large.remaining_allocations: 0
DEBUG: [["total_allocations": 4, "total_allocated_bytes": 528, "remaining_allocations": 0], ["total_allocations": 4, "total_allocated_bytes": 528, "remaining_allocations": 0], ["remaining_allocations": 0, "total_allocated_bytes": 528, "total_allocations": 4], ["total_allocations": 4, "remaining_allocations": 0, "total_allocated_bytes": 528], ["total_allocations": 4, "total_allocated_bytes": 528, "remaining_allocations": 0], ["total_allocated_bytes": 528, "remaining_allocations": 0, "total_allocations": 4], ["total_allocations": 4, "total_allocated_bytes": 528, "remaining_allocations": 0], ["remaining_allocations": 0, "total_allocations": 4, "total_allocated_bytes": 528], ["remaining_allocations": 0, "total_allocated_bytes": 528, "total_allocations": 4], ["total_allocations": 4, "total_allocated_bytes": 528, "remaining_allocations": 0]]
```

while if we did the naive "let's String(describe) the type every time we create keys" (like `            self._storage[ObjectIdentifier(key)] = newValue == nil ? nil : ValueContainer(value: newValue!, string: String(describing: Key.self))` causes:

```
test_002_pass_around_static_strings_large.total_allocated_bytes: 1728
test_002_pass_around_static_strings_large.remaining_allocations: 0
test_002_pass_around_static_strings_large.total_allocations: 28
DEBUG: [["total_allocated_bytes": 1728, "remaining_allocations": 0, "total_allocations": 28], ["total_allocated_bytes": 1728, "remaining_allocations": 0, "total_allocations": 28], ["remaining_allocations": 0, "total_allocations": 28, "total_allocated_bytes": 1728], ["total_allocated_bytes": 1728, "remaining_allocations": 0, "total_allocations": 28], ["remaining_allocations": 0, "total_allocations": 28, "total_allocated_bytes": 1728], ["total_allocated_bytes": 1728, "total_allocations": 28, "remaining_allocations": 0], ["total_allocated_bytes": 1728, "total_allocations": 28, "remaining_allocations": 0], ["remaining_allocations": 0, "total_allocations": 28, "total_allocated_bytes": 1728], ["total_allocations": 28, "total_allocated_bytes": 1728, "remaining_allocations": 0], ["total_allocated_bytes": 1728, "remaining_allocations": 0, "total_allocations": 28]]
```

----

The Swift based benchmark tools allow for:

```
-> % swift run -c release Benchmarks
/Users/ktoso/code/gsoc-swift-tracing/Sources/Benchmarks/ExampleBenchmark.swift:24:9: warning: immutable value 'i' was never used; consider replacing with '_' or removing it
    for i in 1...count {
        ^
        _
[4/4] Linking Benchmarks
#,TEST,SAMPLES,MIN(ns),MAX(ns),MEAN(ns),SD(ns),MEDIAN(ns)
1,ExampleBenchmarks.bench_example,1,0,0,0,0,0

Total performance tests executed: 1
```

Since it optimized the useless loop away, notably in debug mode it performs the needless work:

```
-> % swift run Benchmarks
[3/3] Linking Benchmarks
===========================================================================
=          !!  YOU ARE RUNNING BENCHMARKS IN DEBUG MODE  !!               =
=     When running on the command line, use: `swift run -c release`       =
===========================================================================
#,TEST,SAMPLES,MIN(ns),MAX(ns),MEAN(ns),SD(ns),MEDIAN(ns)
1,ExampleBenchmarks.bench_example,1,49815189,49815189,49815189,0,49815189

Total performance tests executed: 1
```

Pass in `-h` for more information

### CODE REVIEW

File: IntegrationTests/tests_01_performance/test_01_resources/README.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+## Why do I have to set a baseline?
+
+By default this test should always succeed as it doesn't actually compare the number of allocations to a certain number. The reason is that this number varies ever so slightly between operating systems and Swift versions. At the time of writing on macOS we got roughly 326k allocations and on Linux 322k allocations for 1000 HTTP requests & responses. To set a baseline simply run
+
+```bash
+export MAX_ALLOCS_ALLOWED_1000_reqs_1_conn=327000
+```
+
+or similar to set the maximum number of allocations allowed. If the benchmark exceeds these allocations the test will fail.
```
Comment (ktoso) [inline IntegrationTests/tests_01_performance/test_01_resources/README.md:36]: README ;)

File: IntegrationTests/tests_01_performance/test_01_resources/shared.swift [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+enum StringKey1: BaggageContextKey {
+    typealias Value = String
+}
+enum StringKey2: BaggageContextKey {
+    typealias Value = String
+}
+enum StringKey3: BaggageContextKey {
+    typealias Value = String
+}
```
Comment (ktoso) [inline IntegrationTests/tests_01_performance/test_01_resources/shared.swift:43]: shared for all tests

File: IntegrationTests/tests_01_performance/test_01_resources/test_001_pass_around_static_strings_small.swift [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        context[StringKey2.self] = "two"
+        context[StringKey3.self] = "three"
+
+        var numberDone = 1
+        for _ in 0..<1000 {
+            let res = take1(context: context)
+            precondition(res == 42)
+            numberDone += 1
+        }
+        return 1
```
Comment (ktoso) [inline IntegrationTests/tests_01_performance/test_01_resources/test_001_pass_around_static_strings_small.swift]:
```suggestion
        return numberDone
```

File: IntegrationTests/tests_01_performance/test_01_resources/test_001_pass_around_static_strings_small.swift [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    }
+}
```
Comment (ktoso) [inline IntegrationTests/tests_01_performance/test_01_resources/test_001_pass_around_static_strings_small.swift:33]: example for alloc benchmark

File: Sources/Benchmarks/ExampleBenchmark.swift [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    // ...
+}
+
+// completely silly "benchmark" function
+func bench_example(_ count: Int) throws {
+    var sum = 0
+    for i in 1...count {
+        sum += 1
+    }
+}
```
Comment (ktoso) [inline Sources/Benchmarks/ExampleBenchmark.swift:27]: example for swift tools based benchmark; shows howmuch time things take

File: Sources/SwiftBenchmarkTools/ArgParser.swift [hunk 6]
```diff
@@ -0,0 +1,245 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the GSoC Swift Tracing open source project;
+// Based on: https://github.com/apple/swift/tree/cf53143a47278c2a465409a67376642515956777/benchmark/utils
```
Comment (ktoso) [inline Sources/SwiftBenchmarkTools/ArgParser.swift]: Infra based on https://github.com/apple/swift/tree/master/benchmark/utils
It's not super great but good enough to get some basic benches done

File: Package.swift [hunk 7]
```diff
@@ -12,17 +12,64 @@ let package = Package(
...
+            ]
+        ),
 
-        .target(name: "NIOInstrumentation", dependencies: [
-            .product(name: "NIO", package: "swift-nio"),
-            .product(name: "NIOHTTP1", package: "swift-nio"),
-            "Instrumentation"
-        ]),
-        .testTarget(name: "NIOInstrumentationTests", dependencies: ["NIOInstrumentation", "Instrumentation"])
+        // ==== --------------------------------------------------------------------------------------------------------
```
Comment (slashmo) [inline Package.swift:48]: Love it. That's a lot more readable now 👍

**Review by slashmo — APPROVED** [review]

File: Package.swift (hunk 7 above)
Comment (ktoso) [reply Package.swift:48]: Cool, yeah sorry for imposing my usual style here :) Glad you liked it.

**Comment by ktoso** [comment]
CI validation has not run here, weird?

**Review by slashmo — APPROVED** [review]

**Comment by slashmo** [comment]
> CI validation has not run here, weird?

I guess because it originated from your fork, and not from the upstream repo.

**Comment by ktoso** [comment]
> I guess because it originated from your fork, and not from the upstream repo.

That's weird though, shouldn't PR validation run for any PR?

Hm... it passed on master tho https://github.com/slashmo/gsoc-swift-tracing/runs/785192937?check_suite_focus=true so that's ok... but a bit weird, should have run on the PR I'd have hoped?

**Pull request merged**